File: peer-fixtures/textual/sb8_overlay_palette/sb8_overlay_palette_benchmark.py
from __future__ import annotations

import argparse
import asyncio

from textual.app import App, ComposeResult
from textual.widgets import Static


def main() -> None:
    options = parse_args()
    if options.wire:
        WireOverlayApp(options).run()
    else:
        print("Textual SB.8 overlay/palette fixture: use --wire for PTY capture")


class WireOverlayApp(App[None]):
    def __init__(self, options: argparse.Namespace) -> None:
        super().__init__()
        self.options = options
        self.step = 0
        self.query = "open"
        self.palette_open = True

    def compose(self) -> ComposeResult:
        yield Static(self.render_body(), id="body")

    def on_mount(self) -> None:
        self.body = self.query_one("#body", Static)
        asyncio.create_task(self.drive())

    async def drive(self) -> None:
        await asyncio.sleep(self.options.interval_ms / 1000)
        while self.step < self.options.steps:
            self.palette_open = self.step % 3 != 2
            self.query = ["open", "run", "diag", "copy"][self.step % 4]
            self.step += 1
            self.body.update(self.render_body())
            await asyncio.sleep(self.options.interval_ms / 1000)
        self.exit()

    def render_body(self) -> str:
        lines = [
            f"SB.8 overlay churn step={self.step} open={self.palette_open} query={self.query}",
            "",
        ]
        for index in range(9):
            command_index = (self.step + index) % self.options.rows
            lines.append(
                f"screen row {index} focus={index == self.step % 9} command=cmd-{command_index}"
            )
        if self.palette_open:
            lines.extend(["", "+" + "-" * 54 + "+"])
            lines.append(f"| Command Palette query={self.query:<26} |")
            for index in range(6):
                command_index = (self.step * 7 + index) % self.options.rows
                lines.append(f"| cmd-{command_index:04d} {self.query} action-{index:<31} |")
            lines.append("+" + "-" * 54 + "+")
        return "\n".join(lines)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--wire", action="store_true")
    parser.add_argument("--rows", type=positive_int, default=500)
    parser.add_argument("--steps", type=positive_int, default=12)
    parser.add_argument("--interval-ms", type=positive_int, default=40)
    parser.add_argument("--size", default="120x32")
    return parser.parse_args()


def positive_int(value: str) -> int:
    parsed = int(value)
    if parsed <= 0:
        raise argparse.ArgumentTypeError("expected a positive integer")
    return parsed


if __name__ == "__main__":
    main()

File: peer-fixtures/ink/sb8_overlay_palette/bin/sb8_overlay_palette_benchmark.js
#!/usr/bin/env node

import React from 'react'
import { Box, Text, render as renderInk } from 'ink'

const options = parseArgs(process.argv.slice(2))
if (options.wire) await runWire(options)
else console.log('Ink SB.8 overlay/palette fixture: use --wire for PTY capture')

async function runWire(options) {
  let instance
  const done = new Promise((resolve) => {
    instance = renderInk(
      React.createElement(WireOverlayApp, {
        options,
        onDone: () => {
          instance.unmount()
          resolve()
        },
      }),
      {
        exitOnCtrlC: false,
        interactive: true,
        alternateScreen: true,
        maxFps: 60,
        patchConsole: false,
      },
    )
  })
  await done
  await instance.waitUntilExit()
}

function WireOverlayApp({ options, onDone }) {
  const [step, setStep] = React.useState(0)
  const paletteOpen = step % 3 !== 2
  const query = ['open', 'run', 'diag', 'copy'][step % 4]

  React.useEffect(() => {
    const timer = setInterval(() => {
      setStep((current) => {
        if (current + 1 >= options.steps) {
          clearInterval(timer)
          setTimeout(onDone, options.intervalMs)
        }
        return Math.min(current + 1, options.steps)
      })
    }, options.intervalMs)
    return () => clearInterval(timer)
  }, [options.steps, options.intervalMs, onDone])

  return React.createElement(
    Box,
    { flexDirection: 'column' },
    React.createElement(Text, null, `SB.8 overlay churn step=${step} open=${paletteOpen} query=${query}`),
    React.createElement(Text, null, ''),
    ...Array.from({ length: 9 }, (_, index) =>
      React.createElement(
        Text,
        { key: `row-${index}` },
        `screen row ${index} focus=${index === step % 9} command=cmd-${(step + index) % options.rows}`,
      ),
    ),
    paletteOpen
      ? React.createElement(
          React.Fragment,
          null,
          React.createElement(Text, null, ''),
          React.createElement(Text, null, `+${'-'.repeat(54)}+`),
          React.createElement(Text, null, `| Command Palette query=${query.padEnd(26)} |`),
          ...Array.from({ length: 6 }, (_, index) => {
            const commandIndex = (step * 7 + index) % options.rows
            return React.createElement(
              Text,
              { key: `command-${index}` },
              `| cmd-${String(commandIndex).padStart(4, '0')} ${query} action-${String(index).padEnd(31)} |`,
            )
          }),
          React.createElement(Text, null, `+${'-'.repeat(54)}+`),
        )
      : null,
  )
}

function parseArgs(args) {
  const options = { wire: false, rows: 500, steps: 12, intervalMs: 40 }
  for (const arg of args) {
    if (arg === '--wire') options.wire = true
    else if (arg.startsWith('--rows=')) options.rows = positive(arg, '--rows=')
    else if (arg.startsWith('--steps=')) options.steps = positive(arg, '--steps=')
    else if (arg.startsWith('--interval-ms=')) options.intervalMs = positive(arg, '--interval-ms=')
    else if (arg.startsWith('--size=')) continue
    else throw new Error(`unknown argument: ${arg}`)
  }
  return options
}

function positive(arg, prefix) {
  const value = Number(arg.slice(prefix.length))
  if (!Number.isInteger(value) || value <= 0) throw new Error(`${prefix} expects a positive integer`)
  return value
}
